### ISSUE TITLE
aws_ssm_parameter.key_id updated but wasn't changed

### ISSUE THREAD

**Issue by kayman-mk** [hashicorp/terraform-provider-aws#21095]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version
Terraform: 1.0.8
AWS: 3.60

### Affected Resource(s)

* aws_ssm_parameter

### Terraform Configuration Files

```hcl
terraform {
  required_version = ">= 0.13.0"

  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "3.60.0"
    }
  }
}

provider "aws" {
  region = "eu-central-1"
}

resource "aws_ssm_parameter" "foo" {
  name  = "test-foo"
  type  = "String"
  value = "bar"

  key_id = aws_kms_key.foo.arn
}

resource "aws_kms_key" "foo" {
  description             = "TEST"
  deletion_window_in_days = 7
}
```

### Debug Output

https://gist.github.com/kayman-mk/7e8aaa3154d0ddd67f8bd7bd8c431622

### Panic Output

No panic output.

### Expected Behavior

The second `terraform apply` should not change anything.

### Actual Behavior

The aws_ssm_parameter.key_id is always updated by `apply`. In the statefile the key_id field is empty an never written (by the provider?).

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. `terraform apply`

### Important Factoids

Nothing special.

### References

https://github.com/hashicorp/terraform-provider-aws/issues/20150 describes a similar issue with a workaround (use an aws_kms_alias). I also tried it but it didn't work. It seems that the statefile is not written correctly (key_id remains emtpy).

```json
{
      "mode": "managed",
      "type": "aws_ssm_parameter",
      "name": "foo",
      "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "allowed_pattern": "",
            "arn": "arn:aws:ssm:eu-central-1:990563477234:parameter/test-foo",
            "data_type": "text",
            "description": "",
            "id": "test-foo",
            "key_id": "",
            "name": "test-foo",
            "overwrite": null,
            "tags": {},
            "tags_all": {},
            "tier": "Standard",
            "type": "String",
            "value": "bar",
            "version": 19
          },
          "sensitive_attributes": [],
          "private": "bnVsbA==",
          "dependencies": [
            "aws_kms_key.foo"
          ]
        }
      ]
    }
```

**Comment by eugenetaranov** [comment]
I'd like to add that this behaviour is not consistent, I have many resources of type aws_ssm_parameter in my code but only one or two are affected - need for update is detected based on key_id parameter. Resource recreation did not help.

**Comment by svenburkard** [comment]
Guys please check your configured "types". I was also running into this one right now and first thought it's a bug after I found this bug report as a "confirmation" of my own case.

In my case I was just blind and tried to deploy 2 parameters with different types.
One which is just a "**String**" for settings and another one as "**SecureString**" for credentials, which needs to be encrypted.

I configured via "key_id", which key should be used for encryption.
But as I copied my complete resource block and **missed to remove the "key_id" for my normal parameter with "String" type, the "key_id" kept being empty of course, as the "String" type will not encrypt the uploaded parameter.**
For this reason I would have got a change with every tf plan.

I do not know if everyone here was running into the same blindness issue, but I guess there is no bug and at least I see from the details posted by the bug reporter, that he also used the "String" [type](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ssm_parameter#type) in combination with configurations like the "key_id", which is [meant to be used](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ssm_parameter#key_id) together with the "SecureString" instead.

fyi @kayman-mk , @alwaysshrek , @eugenetaranov

**Comment by kayman-mk** [comment]
Yeah, this seems to be the problem. Just checked the Terraform docs and will try with my infrastructure tomorrow. Thanks, @svenburkard

**Comment by kayman-mk** [comment]
You made my day, @svenburkard `key_id` must be used with `type = "SecureString"` only!
